feat(chore): Adding PostsgresEngine class

Itzaprado opened the pull request:
Change Log:

[ ] Addinng PostsgresEngine class
[ ] Adding function to get the IAM email 
[ ] Adding create fiunction to get DB connection
[ ] Adding test file to test DB connection.

ItzayanaPr committed:
feat(engine): add postgresEngine class
- Addinng PostsgresEngine class
- Adding function to get the IAM email
- Adding create fiunction to get DB connection

ItzayanaPr committed:
feat(test): adding engine main test

ItzayanaPr committed:
fix:delete unnecessary file

averikitsch commented:
Please create a directory under "libs/" for this new package with name: "langchain-google-cloud-sql-pg". Please follow the instructions to get this package set up https://js.langchain.com/docs/contributing/integrations. You can use the other provider packages as an example: https://github.com/langchain-ai/langchainjs/tree/main/libs/langchain-google-vertexai

averikitsch commented on langchain-core/package.json:
```diff
@@ -33,14 +33,19 @@
   "author": "LangChain",
   "license": "MIT",
   "dependencies": {
+    "@google-cloud/cloud-sql-connector": "^1.4.0",
```
Please move to the package package.json

averikitsch commented on langchain-core/src/engine.ts:
```diff
@@ -0,0 +1,141 @@
+  static async create(
+      projectId: string,
+      region: string,
+      instanceName: string,
```
If possible, I prefer argument names: `instance` and `database`

Itzaprado replied:
Done, I changed the names

averikitsch commented on langchain-core/src/engine.ts:
```diff
@@ -0,0 +1,141 @@
+
+    const getConnection = async (): Promise<Knex.Knex> => {
+      PostgresEngine.connector = new Connector();
+      const clientOpts = await PostgresEngine.connector.getOptions({
```
IAM type will also have to be set based on user/password or service account, similar to https://github.com/GoogleCloudPlatform/cloud-sql-nodejs-connector/blob/edb840ebe946e8e3a0618703bf5bb58bb015a36b/examples/knex/pg/connect.ts#L35

Itzaprado replied:
I added the validation to use the proper IAM type

averikitsch commented on langchain-core/src/engine.ts:
```diff
@@ -0,0 +1,141 @@
+        connection: {
+          ...clientOpts,
+          user: dbUser,
+          password: password, 
```
Password shouldn't bee needed when using a service account

averikitsch commented on langchain-core/src/engine.ts:
```diff
@@ -0,0 +1,141 @@
+          password: password, 
+          database: dbName,
+        },
+        pool: {
```
Kurtis, should we leave these unset or do we have smart defaults?

averikitsch commented on langchain-core/src/engine.ts:
```diff
@@ -0,0 +1,141 @@
+class PostgresEngine {
+
+  static _createKey = Symbol();
+  private pool: Knex.Knex<any, any[]>;
```
pool should be an instance variable

averikitsch commented on langchain-core/src/engine.ts:
```diff
@@ -0,0 +1,141 @@
+
+class PostgresEngine {
+
+  static _createKey = Symbol();
```
Is this idiomatic to JS?

Itzaprado replied:
Yes, it is a Python-like way of validating creation through the creation method and not through the constructor.

averikitsch commented on langchain-core/src/engine.ts:
```diff
@@ -0,0 +1,141 @@
+   * @returns PostgressEngine instance
+   */
+
+  static async create(
```
We would like this to be called "from_instance"

averikitsch commented on langchain-core/src/engine.ts:
```diff
@@ -0,0 +1,141 @@
+import { Connector, IpAddressTypes} from "@google-cloud/cloud-sql-connector";
+import { GoogleAuth } from "google-auth-library";
+import Knex from "knex";
```
nit: use lower case knex

averikitsch commented on langchain-core/src/engine.ts:
```diff
@@ -0,0 +1,141 @@
+      return Knex(dbConfig)
+    }
+
+    const engine = await getConnection()
```
Does this need to be wrapped in a function or can we just return Knex(dbConfig)?

Itzaprado committed:
feat(chore): moving PostgresEngine class to package
- Addressing comments on PR #1
- Adding test for PostgresEngine instance creation

Itzaprado committed:
fix: remove unnecessary packages and test.

averikitsch commented on libs/langchain-google-cloud-sql-pg/cloud-sql-proxy:
Remove binary. Users will use the built-in connector

averikitsch commented on libs/langchain-google-cloud-sql-pg/package.json:
```diff
@@ -0,0 +1,80 @@
+{
+  "name": "langchain-google-cloud-sql-pg",
```
```suggestion
  "name": "@langchain/langchain-google-cloud-sql-pg",
```

averikitsch commented on libs/langchain-google-cloud-sql-pg/package.json:
```diff
@@ -0,0 +1,80 @@
+{
+  "name": "langchain-google-cloud-sql-pg",
+  "version": "0.0.0",
```
```suggestion
  "version": "0.0.1",
```

averikitsch commented on libs/langchain-google-cloud-sql-pg/package.json:
```diff
@@ -0,0 +1,80 @@
+{
+  "name": "langchain-google-cloud-sql-pg",
+  "version": "0.0.0",
+  "description": "Sample integration for LangChain.js",
```
```suggestion
  "description": "LangChain.js integrations for Google Cloud SQL for PostgreSQL",
```

averikitsch commented on libs/langchain-google-cloud-sql-pg/package.json:
```diff
@@ -0,0 +1,80 @@
+    "type": "git",
+    "url": "git@github.com:langchain-ai/langchainjs.git"
+  },
+  "homepage": "https://github.com/langchain-ai/langchainjs/tree/main/libs/langchain-INTEGRATION_NAME/",
```
Update all "INTEGRATION_NAME" placeholders to "langchain-google-cloud-sql-pg"

averikitsch commented on libs/langchain-google-cloud-sql-pg/package.json:
```diff
@@ -0,0 +1,80 @@
+    "format": "prettier --config .prettierrc --write \"src\"",
+    "format:check": "prettier --config .prettierrc --check \"src\""
+  },
+  "author": "LangChain",
```
```suggestion
  "author": "Google LLC",
```

averikitsch commented on yarn.lock:
```diff
@@ -10245,6 +10245,18 @@ __metadata:
   languageName: node
   linkType: hard
 
+"@google-cloud/cloud-sql-connector@npm:^1.4.0":
```
Can we remove changes to the yarn.lock?

averikitsch commented on libs/langchain-google-cloud-sql-pg/src/engine.ts:
```diff
@@ -0,0 +1,106 @@
+   * @param user Optional - Postgres user name. Defaults to undefined
+   * @param password Optional - Postgres user password. Defaults to undefined
+   * @param iamAccountEmail Optional - IAM service account email. Defaults to undefined
+   * @returns PostgressEngine instance
```
Update all instances of "Postgress" to "Postgres"

averikitsch commented on libs/langchain-google-cloud-sql-pg/src/tests/engine.test.ts:
```diff
@@ -0,0 +1,73 @@
+        process.env.INSTANCE_NAME ?? "",
+        process.env.DB_NAME ?? "",
+        IpAddressTypes.PUBLIC,
+        process.env.DB_USER ?? ""
```
Missing password

averikitsch commented on libs/langchain-google-cloud-sql-pg/src/tests/engine.test.ts:
```diff
@@ -0,0 +1,73 @@
+        process.env.DB_USER ?? ""
+      );
+    }
+    await expect(createInstance).rejects.toBe(
```
Instead of rejects, this should be method "toThrow"

averikitsch commented on libs/langchain-google-cloud-sql-pg/src/engine.ts:
```diff
@@ -0,0 +1,106 @@
+      }
+    }
+
+    PostgresEngine.connector = new Connector();
```
We will need to pass in a user agent for tracking. Can you look at the connector to see if this is possible. I will try to connect internally as well.

Itzaprado replied:
I have been investigating but it seems that the node.js connector does not support the user agent as option parameter.

averikitsch commented on libs/langchain-google-cloud-sql-pg/src/tests/engine.test.ts:
```diff
@@ -0,0 +1,73 @@
+      process.env.INSTANCE_NAME ?? "",
+      process.env.DB_NAME ?? "",
+      IpAddressTypes.PUBLIC, 
+      "",
+      "",
+      process.env.EMAIL ?? "",
```
Forcing the user to pass in undefined values here is not a great experience. It maybe best to support an dictionary for authentication options

Itzaprado committed:
feat(chore): refactoring code and metadata
- Adding dictionary as argument for auth opts
- Change default metadada in package.json
- Remove unnecesary changes and binary

Itzaprado committed:
fix: change validation of user and password

averikitsch reviewed (changes requested):
One last comment on required vs optional args. Thanks!

Itzaprado commented:
> One last comment on required vs optional args. Thanks!

What parameters are you referring to? I have defined these ones https://github.com/averikitsch/langchainjs/pull/1/files#diff-03c57c475cd2d24fb0ff709137c0fc9a920abf9015fdc8354fe9d62545bc703aR12

averikitsch commented on libs/langchain-google-cloud-sql-pg/src/engine.ts:
```diff
@@ -0,0 +1,117 @@
+import { getIAMPrincipalEmail } from "./utils/utils.js";
+
+export interface PostgresEngineArgs {
+  projectId: string,
+  region: string,
+  instance: string,
+  database: string,
```
We should have these as required parameters and use the EngineArgs for the optional IP type, user, password, and iam account

averikitsch commented:
> > One last comment on required vs optional args. Thanks!
> 
> What parameters are you referring to? I have defined these ones https://github.com/averikitsch/langchainjs/pull/1/files#diff-03c57c475cd2d24fb0ff709137c0fc9a920abf9015fdc8354fe9d62545bc703aR12

Sorry, looks like the comment didn't go through! Now added

Itzaprado committed:
feat(chore): change required and optional args in dictionary declaration
